src/kolibri_android/main_activity/activity.py

```diff
@@ -0,0 +1,185 @@
+import logging
+import time
+
+from jnius import autoclass
+
+from ..android_utils import choose_endless_key_uris
+from ..android_utils import get_endless_key_uris
+from ..android_utils import has_any_external_storage_device
+from ..android_utils import PermissionsCancelledError
+from ..android_utils import PermissionsWrongFolderError
+from ..android_utils import provision_endless_key_database
+from ..android_utils import set_endless_key_uris
+from ..android_utils import share_by_intent
+from ..android_utils import StartupState
+from ..android_utils import stat_file
+from ..application import BaseActivity
+from ..kolibri_utils import init_kolibri
+from ..runnable import Runnable
+
+
+PythonActivity = autoclass("org.kivy.android.PythonActivity")
+FullScreen = autoclass("org.learningequality.FullScreen")
+
+
+@Runnable
+def configure_webview(
+    activity, load_runnable, load_with_usb_runnable, loading_ready_runnable
+):
+    FullScreen.configureWebview(
+        activity, load_runnable, load_with_usb_runnable, loading_ready_runnable
+    )
+
+
+@Runnable
+def load_url_in_webview(url):
+    PythonActivity.mWebView.loadUrl(url)
+
+
+@Runnable
+def evaluate_javascript(js_code):
+    PythonActivity.mWebView.evaluateJavascript(js_code, None)
+
+
+def is_endless_key_reachable():
+    """
+    Check if the KOLIBRI_HOME db file is reachable.
+
+    This only works after the user has granted permissions correctly. In other
+    case it always return False.
+    """
+
+    key_uris = get_endless_key_uris()
+    if not key_uris:
+        return False
+    try:
+        # Check if the USB is connected
+        stat_file(key_uris.get("db"))
+        evaluate_javascript("setHasUSB(true)")
+        return True
+    except FileNotFoundError:
+        evaluate_javascript("setHasUSB(false)")
+        return False
+
+
+def wait_until_endless_key_is_reachable():
+    repeat = not is_endless_key_reachable()
+    return repeat
+
+
+def _build_kolibri_process_bus(application):
+    from .kolibri_bus import AppPlugin
+    from .kolibri_bus import KolibriAppProcessBus
+
+    AppPlugin.register_share_file_interface(share_by_intent)
+
+    kolibri_bus = KolibriAppProcessBus(enable_zeroconf=True)
+    AppPlugin(kolibri_bus, application).subscribe()
+
+    return kolibri_bus
+
+
+class MainActivity(BaseActivity):
+    TO_RUN_IN_MAIN = None
+    _last_has_any_check = None
+    _kolibri_bus = None
+
+    def __init__(self):
+        super().__init__()
+
+        configure_webview(
+            PythonActivity.mActivity,
+            Runnable(self._on_load),
+            Runnable(self._on_load_with_usb),
+            Runnable(self._on_loading_ready),
+        )
+
+    def run(self):
+        self.load_url("file:///android_asset/_load.html")
+
+        while True:
```

Comment:
I wish we'll find a better way for the main loop. But it's good keep it as-is for now.

Reply:
Yeah, I agree. I think we can probably set up a message queue or something now that there are some more bones to work with. I figure that overlaps with the other problem where we need to wait for external storage to be available before doing any of it.